net/port_test.go

```diff
@@ -0,0 +1,51 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package net
+
+import "testing"
+
+func TestLocalPortString(t *testing.T) {
+	testCases := []struct {
+		description string
+		ip          string
+		family      IPFamily
+		port        int
+		protocol    string
+		expectedStr string
+	}{
+		{"IPv4 UDP", "1.2.3.4", "", 9999, "udp", "\"IPv4 UDP\" (1.2.3.4:9999/udp)"},
+		{"IPv4 TCP", "5.6.7.8", "", 1053, "tcp", "\"IPv4 TCP\" (5.6.7.8:1053/tcp)"},
+		{"IPv4 TCP", "", IPv4, 1053, "tcp", "\"IPv4 TCP\" (:1053/tcp4)"},
+		{"IPv6 TCP", "2001:db8::1", "", 80, "tcp", "\"IPv6 TCP\" ([2001:db8::1]:80/tcp)"},
+		{"IPv4 SCTP", "9.10.11.12", "", 7777, "sctp", "\"IPv4 SCTP\" (9.10.11.12:7777/sctp)"},
+		{"IPv6 SCTP", "2001:db8::2", "", 80, "sctp", "\"IPv6 SCTP\" ([2001:db8::2]:80/sctp)"},
```

Comment:
is sctp working? I can only see `tcp` and `udp` being handled?

Reply:
there are `sctp` `LocalPort`s being created right now in `kube-proxy` but now that you ask that, as far as I can see they're skipped over. Probably `sctp` support can be removed completely. 
I'm not sure how much to refactor now vs just bringing the code in how it is.

Reply:
I think @danwinship can help us with `sctp` ^^

Reply:
SCTP ports are *intentionally* skipped, per the SCTP KEP. Part of the reason for moving the multiple implementations of this code into one place here is to ensure that everyone gets that behavior and no one accidentally "fixes" it to hold open SCTP ports too.

So these `lp.String()` tests are fine. But also, `OpenLocalPort` should be updated to have:

    case "sctp":
            // SCTP ports are intentionally ignored, to ensure we don't cause the sctp
            // kernel module to be loaded, which breaks userspace SCTP support (and
            // may be considered a security risk by some administrators).
            return nil, nil

and then when kube-proxy and kubelet hostport are updated to use this code, you should be able to remove the checks for SCTP on that end, and just have them call `OpenLocalPort` for all ports, but deal with the fact that they might get back `nil, nil` (ie, no error but also no socket).

Reply:
i was thinking initially that the `LocalPort`s should never be made in the first place for `sctp`.
but updated it now.